packages/wpcom-checkout/src/checkout-line-items.tsx

```diff
@@ -997,8 +1009,24 @@ function WPLineItem( {
 							setIsModalVisible( false );
 						} }
 						primaryAction={ () => {
-							removeProductFromCart( product.uuid ).catch( () => {
-								// Nothing needs to be done here. CartMessages will display the error to the user.
+							let product_uuids_to_remove = [ product.uuid ];
+
+							// Gifts need to be all or nothing, to prevent leaving
+							// the site in a state where it requires other purchases
+							// in order to actually work correctly for the period of
+							// the gift (for example, gifting a plan renewal without
+							// a domain renewal would likely lead the site's domain
+							// to expire soon afterwards).
+							if ( product.is_gift_purchase ) {
+								product_uuids_to_remove = responseCart.products
+									.filter( ( cart_product ) => cart_product.is_gift_purchase )
+									.map( ( cart_product ) => cart_product.uuid );
+							}
+
+							product_uuids_to_remove.forEach( function ( product_uuid ) {
```

Comment:
Alternatively, we could use `replaceProductsInCart( responseCart.products.filter( product => ! product.is_gift_purchase ) )` if you want to keep non-gifts.

Reply:
Ohh. I didn't look at the context of where we were. Both `replaceProductsInCart` and `removeProductFromCart` are provided by the same hook, `useShoppingCart`, but I didn't notice that here we're not calling that hook. `removeProductFromCart` is being passed-in. The reason is because we're augmenting the function 

https://github.com/Automattic/wp-calypso/blob/ea0380a9ff947d5dae7c61338f197e7f2741d9ff/client/my-sites/checkout/composite-checkout/components/checkout-main.tsx#L309-L314

so that it's actually calling `removeProductFromCartAndMaybeRedirect`. Since they all use the same Promises, it should be fine to call that multiple times in this function.

In that case, let's just use `removeProductFromCart` as you suggested.